components/brave_service_keys/service_key_utils.cc

```diff
@@ -0,0 +1,104 @@
+
+#include <vector>
+
+#include "base/base64.h"
+#include "base/logging.h"
+#include "base/strings/strcat.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_service_keys/buildflags.h"
+#include "crypto/hmac.h"
+#include "crypto/sha2.h"
+
+namespace brave_service_keys {
+
+namespace {
+
+constexpr char kAuthorization[] = "authorization";
+constexpr char kDigest[] = "digest";
+constexpr char kRequestTarget[] = "(request-target)";
+
+}  // namespace
+
+std::pair<std::string, std::string> GetDigestHeader(
+    const std::string& payload) {
+  const std::string value = base::StrCat(
+      {"SHA-256=", base::Base64Encode(crypto::SHA256HashString(payload))});
+  return std::make_pair(kDigest, value);
+}
+
+std::pair<std::string, std::string> CreateSignatureString(
+    base::flat_map<std::string, std::string> headers,
+    const GURL& url,
+    const std::string& method,
+    const std::vector<std::string>& headers_to_sign) {
+  // Create and add the (request-target) header if included
+  if (std::find(headers_to_sign.begin(), headers_to_sign.end(),
+                kRequestTarget) != headers_to_sign.end()) {
+    headers.insert(std::make_pair(
+        kRequestTarget,
+        base::StrCat({base::ToLowerASCII(method), " ", url.PathForRequest()})));
+  }
+
+  std::string header_names;
+  std::string signature_string;
+
+  for (const auto& key : headers_to_sign) {
+    auto header = headers.find(key);
+    if (header != headers.end()) {
+      if (!header_names.empty()) {
+        header_names += " ";
+        signature_string += "\n";
+      }
+      header_names += key;
+      signature_string += key + ": " + header->second;
+    } else {
+      VLOG(1) << __func__ << "Can't sign over non existant header " << key;
+      DCHECK(false);
+    }
+  }
+
+  return {header_names, signature_string};
+}
+
+std::optional<std::pair<std::string, std::string>> GetAuthorizationHeader(
+    const std::string& service_key,
+    const base::flat_map<std::string, std::string>& headers,
+    const GURL& url,
+    const std::string& method,
+    const std::vector<std::string>& headers_to_sign) {
+  auto [header_names, signature_string] =
+      CreateSignatureString(headers, url, method, headers_to_sign);
+
+  // Create the signature using the service_key.
+  crypto::HMAC hmac(crypto::HMAC::SHA256);
+  const size_t signature_digest_length = hmac.DigestLength();
+  std::vector<uint8_t> signature_digest(signature_digest_length);
+  const bool success = hmac.Init(service_key) &&
+                       hmac.Sign(signature_string, &signature_digest[0],
+                                 signature_digest.size());
+  if (!success) {
+    return std::nullopt;
+  }
+
+  // Create the authorization header.
+  std::string signature_digest_base64;
+  base::Base64Encode(
+      std::string(signature_digest.begin(), signature_digest.end()),
+      &signature_digest_base64);
+
+  const std::string value = base::StrCat(
+      {"Signature "
```

Comment:
the formatting here is very strange, please clang format this. Also you're missing a comma